### PR TITLE
Add ability to set a contact method your default contact method

### PR DESCRIPTION
**Pull request by Beamanator**
@aimane-chnaif I'm sorry but you probably won't be able to complete these tests till https://github.com/Expensify/Web-Expensify/pull/36920 gets to staging

Also cc @cristipaval since you've been helping work on similar pages for a while now 👍 

### Details

Adds one more piece of functionality for contact methods: Being able to set them as default.

Note: In OldDot, we have some domain security settings that disable users from changing their default login. However, none of the domain functionality is on NewDot (yet) so we are only pulling in the user's domain security group data so we can make sure the user knows if they are or aren't able to update their primary login (default contact method).

### Fixed Issues
$ https://github.com/Expensify/App/issues/16747

### Tests
1. ~If https://github.com/Expensify/Web-Expensify/pull/36920 isn't merged yet, check it out to start testing~
2. ~Open up OldDot, sign in with a domain admin account~
    - ~Note: Your account should be on the `passwordless` beta (or if you're an Expensify employee, you should already have access to all betas)~
    - ~You can probably fake this by returning `true` in `Permissions.canUsePasswordlessLogins`~
    - Note: I removed this point b/c passwordless is live!
3. Go to domain settings (https://www.expensify.com.dev/domain_groups), check out the security group settings for that account you're signed in as. Make sure "Restrict primary login selection" is false
    - Screenshot 2023-04-06 at 1 29 49 PM
4. Log in to NewDot with that same account, go to http://localhost:8080/settings/profile/contact-methods
5. Add a secondary login and verify it
6. Now open that contact method's "details" page, and verify you see the "Set as default" button
    - Screenshot 2023-04-06 at 1 31 22 PM
7. Click the button, verify your secondary login was correctly set as default.
    - You should now see "This is your current default contact method. You will not be able to delete this contact method until you set an alternative default by selecting another contact method and pressing “Set as default”." on that contact's details page
    - Screenshot 2023-04-06 at 3 09 46 PM
    - Screenshot 2023-04-06 at 3 09 30 PM
8. Change back to your other login as default
9. Next, back in OldDot domain settings, change "Restrict primary login selection" to true
    - Screenshot 2023-04-06 at 3 10 47 PM
10. Back on http://localhost:8080/settings/profile/contact-methods, open up your secondary login's "details" page
    - Note: You shouldn't have to refresh, because recently we made Onyx update when domain security group settings change in OldDot!
12. Verify you now only see the "Delete" option
    - Screenshot 2023-04-06 at 3 11 39 PM

Final test to make sure personal details update correctly:
1. Create a brand new account on a public domain (ex: `...@gmail.com`) and DON'T set a display name
2. Add a new contact method for this account
3. Set the new contact method as default
4. Close the sidebar, open settings
5. Verify the settings page shows all of the correct data immediately
6. Navigate to other pages of the app, send messages, etc - making sure nothing is broken and that no console errors show up

**NOTE:** Your old default contact method will still show above messages you sent in chat reports (assuming you don't have a display name set). This is because the "old default contact method" was sent from the server, and we don't update ALL report actions any time a default contact method is changed.
- I plan to create a follow-up for this, but we already discussed it's find to most past this for now

- [ ] Verify that no errors appear in the JS console

### Offline tests
1. Switch your security group to not have primary login switching disabled again (same as step 3 above)
8. Refresh NewDot and open up the details page of your secondary login you've been testing with
9. Go offline, click "Set as default"
10. Verify you see the button is now grayed out, like this:
    - Screenshot 2023-04-06 at 3 33 29 PM

### QA Steps
Same as above

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/3885503/230400750-ff0f6927-4294-4c16-bc8c-2a589531097b.mov

Domain setting change test:

https://user-images.githubusercontent.com/3885503/233671700-bb141e3f-429c-4cf8-bd29-91fcd10dac5f.mov

</details>

<details>
<summary>Mobile Web - Chrome</summary>

Domain setting change test:

https://user-images.githubusercontent.com/3885503/233676876-81551abd-dc4e-4666-af32-7ac1048793fc.mov

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/3885503/234012021-550ae98b-f9d5-4dbc-9bbe-e84de91a1794.mov

</details>

<details>
<summary>Desktop</summary>

Domain setting change test:

https://user-images.githubusercontent.com/3885503/233673669-cf2a3098-7f5e-4895-9db7-870c0c443835.mov

</details>

<details>
<summary>iOS</summary>

Domain setting change test:

https://user-images.githubusercontent.com/3885503/234010127-54149236-4270-4fee-8857-d248e4948b74.mov

</details>

<details>
<summary>Android</summary>

Domain setting change test:

https://user-images.githubusercontent.com/3885503/233676459-fe1cb4d6-562e-470d-9f62-7790a4998403.mov

</details>

### CODE REVIEW

**Comment by Beamanator** [comment]
FYI I haven't completed the videos on all platforms yet, mainly because my wifi is MISERABLE today and it's really slowing down lots of things 🙃

File: src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js
Comment (deetergp) [inline src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js:132]: NAB: You could just `return lodashGet(…)` directly without assigning the variable.

**Review by deetergp — DISMISSED** [review]
Code looks good, despite my one NAB comment. I did check out the Web-E PR and tested it in macOS Chrome and it worked as described 👍

File: src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js
Comment (Beamanator) [reply src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js]: Done 👍

**Comment by Beamanator** [comment]
One more point I'm thinking about with this PR: How to show the Password textinput for users not on the passwordless beta

Internal slack convo [here](https://expensify.slack.com/archives/C048ZLJ78P5/p1680876428606049)

File: src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js
Comment (Beamanator) [reply src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js:191]: Aahhhh good call I've seen that and I forgot about it 👍

**Comment by Beamanator** [comment]
@aimane-chnaif 

> Not related to PR, but I am a bit confused why it says "Phone number" to add new contact method, while both email and phone number are acceptable.

Yeah good question, don't worry too much about that since that's the `LoginField` which @cristipaval is removing very very soon in this PR: https://github.com/Expensify/App/pull/15330



> Though I am not in passwordless beta, it still shows magic code input. This is not handled yet, right?

Correct, it's noted in this polish issue for multiple contact methods-related updates: https://github.com/Expensify/App/issues/16666

**Comment by Beamanator** [comment]
Waiting on confirmation from one [question](https://expensify.slack.com/archives/C048ZLJ78P5/p1681206216368619?thread_ts=1680876428.606049&cid=C048ZLJ78P5) - basically we've gone back 'n forth on showing or not showing a password field for "Set as default" functionality (for non-passwordless users).

**Comment by Beamanator** [comment]
> Waiting on confirmation from one [question](https://expensify.slack.com/archives/C048ZLJ78P5/p1681206216368619?thread_ts=1680876428.606049&cid=C048ZLJ78P5) - basically we've gone back 'n forth on showing or not showing a password field for "Set as default" functionality (for non-passwordless users).

Ok so the plan now is to only show "Set as default" functionality for passwordless users so that we don't need the password input at all on the contact method details page 👍

**Comment by Beamanator** [comment]
@aimane-chnaif @deetergp @cristipaval  I believe this PR is ready for review!!

Note: I didn't finish all the videos, sorry about that - I'm mainly OOO this week but I'd love if we can get this merged by EOW, so I'll try to jump on to respond to any questions / bugs you may find 😬

File: src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js
Comment (cristipaval) [inline src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js:219]: 👌 love the detailed comment

**Review by cristipaval — DISMISSED** [review]
Looking great! @aimane-chnaif , Could you review and test? I will pull the main branch into this to resolve the conflicts and prepare it since @Beamanator is ooo.

**Comment by cristipaval** [comment]
Alright @aimane-chnaif , it's ready for review. Please prioritize this tomorrow, we want to get **Contact Methods** out the door.

**Comment by Beamanator** [comment]
Bump @aimane-chnaif - any chance you can get to this early this week? If not maybe need to assign a new C+

**Comment by Beamanator** [comment]
@aimane-chnaif yes agreed - feel free to mention them here but I'd probably recommend posting them in Slack too and tagging me and / or @cristipaval

**Comment by cristipaval** [comment]
@aimane-chnaif I enabled passwordless beta for your accounts.

**Comment by Beamanator** [comment]
> When both accounts are passwordless, this error happens. Is this expected?
I created 2 new accounts: [aiman.chnaif+beta1@gmail.com](mailto:aiman.chnaif+beta1@gmail.com), [aiman.chnaif+beta2@gmail.com](mailto:aiman.chnaif+beta2@gmail.com) and you made them passwordless. No other actions I did.
I tried to add [aiman.chnaif+beta2@gmail.com](mailto:aiman.chnaif+beta2@gmail.com) as default method in [aiman.chnaif+beta1@gmail.com](mailto:aiman.chnaif+beta1@gmail.com) account.

@aimane-chnaif yeah in this case, if both `aiman.chnaif+beta1@gmail.com` and `aiman.chnaif+beta2@gmail.com` are already completely unique accounts, you won't be able to add either as a 2ndary login, you'll have to merge them first. I'm not 100% sure where we're at with merging accounts in NewDot, I'll ask in Slack 👍

File: src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js
Comment (Beamanator) [reply src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js:218]: Great question, you'd have to have an account on some domain (ex: expensify.com, stars-egypt.org, costacoffee.com) - not a public domain (ex: gmail.com, yahoo.com, etc)

Comment (Beamanator) [reply src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js:218]: Do you have an account like this? Maybe some old school or work email account?

Comment (Beamanator) [reply src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js:218]: Sadly hotmail also won't work, you have to use an email from some organization... If you don't have one that's fine, we can test that flow 👍

**Comment by Beamanator** [comment]
Ok so just so we're all on the same page:
1. We don't need to test the passwordless / non-passwordless combination accounts since all accounts will be passwordless very soon 💪 
2. @aimane-chnaif doesn't have an account that's not on a public domain, so @deetergp @cristipaval @Beamanator one of us will have to test this flow (specifically, editing domain security group settings to allow / disallow the changing of primary logins)
3. @aimane-chnaif are you able to at least test the offline case & post videos of that & successfully setting a contact method as default on a public domain account? (a.k.a. an email address w/ ...@gmail.com or a phone number account)

**Comment by Beamanator** [comment]
Ok finished & uploaded the test (changing domain settings & seeing the update reflected in EChat) on all platforms except iOS, will try to do that tonight

**Review by cristipaval — DISMISSED** [review]

**Comment by Beamanator** [comment]
Uploaded videos on all platforms for changing domain security groups & seeing the updates in NewDot 👍 

Thanks for the approval @cristipaval ! 👍 

@aimane-chnaif are you able to finish testing today or no?

**Comment by Beamanator** [comment]
@aimane-chnaif thanks for noting what happens in that flow! @cristipaval and I agreed that is something we should handle _eventually_ but is a super edge case so we'll add it to the list of contact method polish items ([here](https://github.com/Expensify/App/issues/16666)). To add to this point, I believe that case could also happen if you have 2 clients logged in at the same time and you set the same contact method as default on both devices at the same time. AGain, this is an extreme edge case so we don't care to handle that now, but we'll handle as a follow-up.

Please proceed with testing, thanks 🙏 👍

**Comment by Beamanator** [comment]
> @Beamanator when add new email (not signed up Expensify yet) as secondary contact and then set it as default, what's the expected behavior?

@aimane-chnaif  I'm not exactly sure what you mean - a secondary login / contact method is just another way for you to log in to the same account

**Comment by Beamanator** [comment]
Hmm this is being hit: https://github.com/Expensify/App/blob/7d23b518b953c6ea2caa7ea812b874c18bb4be5d/src/pages/settings/InitialSettingsPage.js#L270-L272

**Comment by Beamanator** [comment]
Ohhhhhh the user's new contact method is not set in `personalDetails` yet, which is why that returns `undefined` 👍

**Comment by Beamanator** [comment]
@aimane-chnaif great catch on that case, hopefully it's fixed now 🙏  Please test again when you get a chance!

**Comment by Beamanator** [comment]
@aimane-chnaif I added this test to cover the new case we just solved:

> Final test to make sure personal details update correctly:
1. Create a brand new account on a public domain (ex: `...@gmail.com`)
2. Add a new contact method for this account
3. Set the new contact method as default
4. Close the sidebar, open settings
5. Verify the settings page shows all of the correct data immediately
6. Navigate to other pages of the app, send messages, etc - making sure nothing is broken and that no console errors show up

**Comment by Beamanator** [comment]
@aimane-chnaif I'm back to bump this 😬 Have you had a chance to test that flow again?

**Comment by Beamanator** [comment]
Thanks @aimane-chnaif - I will try to test that out soon 🙏

**Comment by Beamanator** [comment]
Another interesting thing happening is: When we get report actions (I assume using `OpenReport`), we send an `actorEmail` that matches the user's current default login. When the default login changes, report actions don't update (with the new user login) - but if we refresh they get updated.

So we either need to:
1. Optimistically change `actorEmail` in ALL reports & ALL actions sent by the logged in user, or
2. Figure out how to use the session email for the logged-in user

Note: In the future we "may" want to tell other users when your default login email changes. Iff we end up doing that, we'll run into the same issue for having to update all of the reportActions in other chat reports 🤔  I'll probably have to start a quick discussion in slack to get some ideas what we prefer here

**Comment by Beamanator** [comment]
Ok I asked for ideas / thoughts on the solutions here: https://expensify.slack.com/archives/C01GTK53T8Q/p1682697131052759

**Comment by Beamanator** [comment]
@aimane-chnaif yes thanks will do, I think I may have to put this on hold for a bit, but I'm planning to start a discussion in Slack soon

**Comment by Beamanator** [comment]
Alright in slack [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1683282627072939) I mentioned it would be nice to move forward without dealing with the updating of report actions (when a contact method is set as default) because:
1. This would be a very large undertaking for this PR
2. There's already a planned similar change (re-keying personal details by accountID)
3. We don't want to delay this PR any longer b/c this is the last major change for the Account Settings project!

**Comment by Beamanator** [comment]
So I'm working on pulling main & retesting, then I'll ask @aimane-chnaif to do the same before we move to hopefully get this merged soon!

File: src/libs/actions/Session/index.js
Comment (Beamanator) [inline src/libs/actions/Session/index.js:1]: FYI these changes are due to this bug: https://github.com/Expensify/App/pull/15811#discussion_r1187244419

This might be in the process of updating somewhere else but I thought I might as well fix it here

**Comment by Beamanator** [comment]
Ok @aimane-chnaif I tested again and everything looks like it's working well 👍 

Can you & @cristipaval please test when you have a chance? (I know @aimane-chnaif you can't test the domain security group changes, so hopefully @cristipaval / @deetergp can test that 🙏 )

**Comment by Beamanator** [comment]
Thanks @aimane-chnaif - yeah notifications are super mucked up right now haha

**Comment by Beamanator** [comment]
Good point @aimane-chnaif - I made the change to only update the current client instantly, I will definitely add that as a follow-up to make sure other devices get the same update 👍

**Review by aimane-chnaif — DISMISSED** [review]
Let's fix above issues as follow-up, not marking as blocker.
Other than that, tests well.

**Comment by Beamanator** [comment]
Thanks for reviewing @aimane-chnaif !

> Another bug:
> 
> * X set profile avatar A
> * change default contact to Y
> * Y now has same avatar as X (A)
> * on Y, set profile avatar B
> * change default contact back to X
> * X's avatar changed to B
> 
> So my question is one profile avatar is shared between contacts?

It sounds like this is expected - the user should only ever have 1 profile avatar. If you change your default contact method, the avatar should stay the same. That's what's happening, right? Just making sure I followed correctly

File: src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js
Comment (deetergp) [inline src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js:231]: Why aren't `!isDefaultContactMethod` and `loginData.validatedDate` just not conditions included inside `getCanChangeDefaultContactMethod()`? Instead of a new const, why not just rename the method to `canChangeDefaultContactMethod()` and call it directly?

**Review by deetergp — CHANGES_REQUESTED** [review]

File: src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js
Comment (Beamanator) [reply src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js:231]: Iiiiii think that would be clearer, good call both of y'all :D

**Comment by Beamanator** [comment]
I'm just about to jump into fixing merge conflicts & the requested change 👍

**Comment by Beamanator** [comment]
- Merge conflict fixed & updated to make sure prettier is happy
- Updated the function `canChangeDefaultContactMethod` to do all logic regarding if a user can change their default contact method or not 👍 

Ready for re-review @deetergp @aimane-chnaif 🙏

**Review by deetergp — APPROVED** [review]
Looking good! Thanks for the changes. 👍

**Review by aimane-chnaif — APPROVED** [review]
LGTM2

**Comment by Beamanator** [comment]
Nice! @deetergp @cristipaval either of you want the merge honors? :D

**Review by cristipaval — APPROVED** [review]

**Pull request merged**
